### PR TITLE
Feat: Add Billable Toggle UI

### PR DESCRIPTION
**Pull request by waterim**
### Details

This PR is about to add Billable toggle to money request.
To test please insert this to App.js with your policyID you are testing:
```
    window.Onyx.merge(`policy_${YOUR_ID}`, {
        disabledFields: {
            defaultBillable: false,
            reimbursable: false,
        },
        defaultBillable: false,
    });
```
### Fixed Issues

$ https://github.com/Expensify/App/issues/26128
PROPOSAL: https://github.com/Expensify/App/issues/26128#issuecomment-1699037140

### Tests

1. Login to the app
2. Click "+" dropdown
3. Click on "Request money"
4. Type any number
5. Click "Next"
6. Select any email workspace
7. Click "Show more"
8. Look at the toggler
9. Default state of the toggle should have based on a policy's settings set in OldDot
10. After we refresh page, selected toggle should remain 

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as tests

### QA Steps

None

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [x] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Web</summary>

image
image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

image

image

</details>

<details>
<summary>Mobile Web - Safari</summary>

image
image

</details>

<details>
<summary>Desktop</summary>

image
image

</details>

<details>
<summary>iOS</summary>

image
image

</details>

<details>
<summary>Android</summary>

image

image

</details>

### CODE REVIEW

File: src/components/MoneyRequestConfirmationList.js [hunk 1]
```diff
@@ -499,6 +501,18 @@ function MoneyRequestConfirmationList(props) {
                             disabled={didConfirm || props.isReadOnly}
                         />
                     )}
+                    <View>
```
Comment (koko57) [inline src/components/MoneyRequestConfirmationList.js:504]: do we need this additional view here?

File: src/components/MoneyRequestConfirmationList.js [hunk 2]
```diff
@@ -499,6 +501,18 @@ function MoneyRequestConfirmationList(props) {
...
+                        <View style={[styles.flexRow, styles.mb4, styles.justifyContentBetween, styles.alignItemsCenter, styles.ml5, styles.mr8]}>
+                            <Text color={!props.iouIsBillable ? themeColors.textSupporting : undefined}>{translate('common.billable')}</Text>
+                            <Switch
+                                accessibilityLabel={translate('common.billable')}
+                                isOn={props.iouIsBillable}
+                                onToggle={(val) => {
+                                    props.onToggleBillable(val);
+                                }}
```
Comment (thiagobrez) [inline src/components/MoneyRequestConfirmationList.js:512]: We could simplify here:

```suggestion
                                onToggle={props.onToggleBillable}
```

File: src/pages/iou/steps/MoneyRequestConfirmPage.js [hunk 3]
```diff
@@ -283,6 +284,8 @@ function MoneyRequestConfirmPage(props) {
                                 iouAmount={props.iou.amount}
                                 iouComment={props.iou.comment}
                                 iouCurrencyCode={props.iou.currency}
+                                iouIsBillable={isBillable}
+                                onToggleBillable={(newValue) => setIsBillable(newValue)}
```
Comment (thiagobrez) [inline src/pages/iou/steps/MoneyRequestConfirmPage.js:288]: Same here:

```suggestion
                                onToggleBillable={setIsBillable}
```

File: src/pages/iou/steps/MoneyRequestConfirmPage.js [hunk 4]
```diff
@@ -59,6 +59,7 @@ const defaultProps = {
 };
 
 function MoneyRequestConfirmPage(props) {
+    const [isBillable, setIsBillable] = useState(props.policy?.defaultBillable);
```
Comment (rezkiy37) [inline src/pages/iou/steps/MoneyRequestConfirmPage.js:62]:
```suggestion
    const [isBillable, setIsBillable] = useState(props.policy?.defaultBillable ?? false);
```

Comment (pac-guerreiro) [reply src/pages/iou/steps/MoneyRequestConfirmPage.js:62]: @rezkiy37 `const [isBillable, setIsBillable] = useState(!!props.policy?.defaultBillable);` 🤯

**Comment by waterim** [comment]
Just have one stupid question, do I need to send "defaultBillable" or another name?
@amyevans

**Comment by amyevans** [comment]
@waterim Do you mean when sending to the API? Let's use `billable`

**Comment by waterim** [comment]
@amyevans Okay, will change it, thank you

**Comment by puneetlath** [comment]
@waterim looks like you're missing screenshots/videos.

**Comment by puneetlath** [comment]
Right now I don't think it's tied to any beta, but this also isn't being sent by the back-end yet, so you'll need to add the fields in Onyx like @waterim recommended in the OP. 

@amyevans should this be under the tags beta? Or is it fine to just have it as-is.

**Comment by amyevans** [comment]
We should put it under the tags beta since we're trying to work incrementally so it's not a complete feature yet (and thus we don't want contributors reporting "bugs"). @waterim can you adjust your PR to check for the tags beta before displaying any UI?

**Comment by waterim** [comment]
@amyevans Sorry, thats because of the conflicts resolve, will change now

**Comment by puneetlath** [comment]
@waterim just to make sure we're on the same page. You'll want to add the `canUseTags` check that we added here: https://github.com/Expensify/App/pull/26954/files

If the user has the permission, then we'll want to show the billable toggle.

**Comment by waterim** [comment]
@puneetlath ready for

**Comment by waterim** [comment]
@0xmiroslav done

File: src/components/MoneyRequestConfirmationList.js [hunk 5]
```diff
@@ -530,6 +532,16 @@ function MoneyRequestConfirmationList(props) {
                             disabled={didConfirm || props.isReadOnly}
                         />
                     )}
+                    {canUseTags && !lodashGet(props.policy, 'disabledFields.defaultBillable', true) && (
```
[the remark this answers is not in the document]
Comment (waterim) [reply src/components/MoneyRequestConfirmationList.js:535]: @amyevans to be sure

[the remark this answers is not in the document]
Comment (puneetlath) [reply src/components/MoneyRequestConfirmationList.js:535]: Yes, that's correct.

[the remark this answers is not in the document]
Comment (amyevans) [reply src/components/MoneyRequestConfirmationList.js:535]: Yes, correct. An IOU request doesn't have an associated policy ID so the field is never going to show for one, which is intended.

[the remark this answers is not in the document]
Comment (amyevans) [reply src/components/MoneyRequestConfirmationList.js:535]: Whoops I was too slow. Thanks Puneet 😄

**Comment by waterim** [comment]
@0xmiroslav Updated my solution, removed state, now after refresh it takes data from onyx(as all other fields), but after reset(for example go out from the request) it will be updated with default data from policy as it should be.

File: src/libs/actions/IOU.js [hunk 6]
```diff
@@ -95,7 +95,7 @@ function resetMoneyRequestInfo(id = '') {
         receiptPath: '',
         receiptSource: '',
         transactionID: '',
-        defaultBillable: false,
+        defaultBillable: '',
```
[the remark this answers is not in the document]
Comment (waterim) [reply src/libs/actions/IOU.js:98]: To make it empty, but agree, updated with null

**Review by 0xmiroslav — APPROVED** [review]
LGTM!

Left minor feedback but NAB

File: src/pages/iou/steps/MoneyRequestConfirmPage.js [hunk 7]
```diff
@@ -77,6 +77,7 @@ function MoneyRequestConfirmPage(props) {
         if (policyExpenseChat) {
             Policy.openDraftWorkspaceRequest(policyExpenseChat.policyID);
         }
+        if (typeof props.iou.defaultBillable !== 'boolean') IOU.setMoneyRequestDefaultBillable(lodashGet(props.policy, 'defaultBillable', false));
```
[the remark this answers is not in the document]
Comment (amyevans) [reply src/pages/iou/steps/MoneyRequestConfirmPage.js:80]: +1 to the code style request, and also please add a comment explaining why we're doing this

**Review by amyevans — CHANGES_REQUESTED** [review]
A few small comments! Also, the Tests section and QA section need updating. For QA, let's say "None" because we're still behind a beta and don't want deploy blockers created for any failures. For tests:

1. Login to the app
2. Click "+" dropdown
3. Click on "Request money"
4. Type any number
5. Click "Next"
6. Select any ~email~ workspace
7. Click "Show more"
8. Look at the toggler
9. _Please describe what default state the toggle should have based on a policy's settings set in OldDot_
10. _Please describe what should happen if you refresh the page_

**Comment by waterim** [comment]
Resolved @amyevans

**Review by amyevans — COMMENTED** [review]
Just 1 super minor update, then we can :shipit:

File: src/libs/actions/IOU.js [hunk 8]
```diff
@@ -338,6 +339,7 @@ function buildOnyxDataForMoneyRequest(
  * @param {Object} [receipt]
  * @param {String} [existingTransactionID]
  * @param {String} [category]
+ * @param {Boolean} billable
```
[the remark this answers is not in the document]
Comment (waterim) [reply src/libs/actions/IOU.js]: resolved

**Review by amyevans — APPROVED** [review]
Awesome 😎

File: src/pages/iou/steps/MoneyRequestConfirmPage.js [hunk 9]
```diff
@@ -352,4 +360,9 @@ export default compose(
             key: `${ONYXKEYS.COLLECTION.SELECTED_TAB}${CONST.TAB.RECEIPT_TAB_ID}`,
         },
     }),
+    withOnyx({
+        policy: {
+            key: ({report}) => `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`,
+        },
+    }),
```
Comment (s77rt) [inline src/pages/iou/steps/MoneyRequestConfirmPage.js:367]: This caused a regression https://github.com/Expensify/App/issues/28182. The report object may be undefined and accessing the policyID on an undefined crashes the app.

File: src/components/MoneyRequestConfirmationList.js [hunk 10]
```diff
@@ -530,6 +532,16 @@ function MoneyRequestConfirmationList(props) {
...
+                        <View style={[styles.flexRow, styles.mb4, styles.justifyContentBetween, styles.alignItemsCenter, styles.ml5, styles.mr8]}>
+                            <Text color={!props.iouIsBillable ? themeColors.textSupporting : undefined}>{translate('common.billable')}</Text>
+                            <Switch
+                                accessibilityLabel={translate('common.billable')}
+                                isOn={props.iouIsBillable}
+                                onToggle={props.onToggleBillable}
```
Comment (tgolen) [inline src/components/MoneyRequestConfirmationList.js:541]: It appears this was never added to the propType definitions. Can you please add it?

Comment (amyevans) [reply src/components/MoneyRequestConfirmationList.js:541]: In progress here: https://github.com/Expensify/App/pull/29327

Thanks for flagging!

**Pull request merged**